tracing-fmt/src/writer.rs

```diff
@@ -0,0 +1,165 @@
+//! Abstractions for creating [`io::Write`] instances.
+
+use super::Builder;
+use std::io;
+
+/// A type that can create [`io::Write`] instances.
+///
+/// `NewWriter` is used by [`FmtSubscriber`] to print formatted text representations of [`Event`]s.
+///
+/// This trait is already implemented for function pointers and immutably-borrowing closures that
+/// return an instance of [`io::Write`], such as [`io::stdout`] and [`io::stderr`].
+///
+/// [`FmtSubscriber`]: crate::FmtSubscriber
+/// [`Event`]: tracing_core::Event
+pub trait NewWriter {
+    /// The concrete [`io::Write`] implementation returned by [`new_writer`].
+    ///
+    /// [`new_writer`]: NewWriter::new_writer
+    type Writer: io::Write;
+
+    /// Returns an instance of [`Writer`].
+    ///
+    /// # Implementer notes
+    ///
+    /// [`FmtSubscriber`] will call this method each time an event is recorded. Ensure any state
+    /// that must be saved across writes is not lost when the [`Writer`] instance is dropped. If
+    /// creating a [`io::Write`] instance is expensive, be sure to cache it when implementing
+    /// [`NewWriter`] to improve performance.
+    ///
+    /// [`Writer`]: NewWriter::Writer
+    /// [`FmtSubscriber`]: crate::FmtSubscriber
+    fn new_writer(&self) -> Self::Writer;
+}
+
+impl<F, W> NewWriter for F
+where
+    F: Fn() -> W,
+    W: io::Write,
+{
+    type Writer = W;
+
+    fn new_writer(&self) -> Self::Writer {
+        (self)()
+    }
+}
+
+impl<N, E, F, W> Builder<N, E, F, W> {
```

Comment:
Nit: IMO, it is a little strange that this impl block is in the `writer` module, rather than in `lib.rs` with all the other methods on `Builder`. You'll note that although the formatting traits are defined in their own module, the `Builder` methods that configure them are still with the rest of the methods on that type.

Can we make this more consistent by moving this to the `impl Builder` block in lib.rs? Not a big deal though.

Reply:
I can put it back for this PR, but I'm wondering if in the long run it may be better to have these methods in their respective sub-modules, especially if you plan on adding additional builder configuration later on.

Reply:
I could go either way. IMO it's easier to make changes when you can find all the methods in one place, and leads to a cleaner diff (otherwise, adding a new field to the builder would change a lot of files that might not otherwise be related to the change). However, I'm willing to be convinced otherwise.

All that really matters to me is that this is consistent, so if we're going to change this, I'd prefer we change everything in one go.